dbms/src/Dictionaries/CacheDictionary.h

```diff
@@ -284,15 +317,135 @@ class CacheDictionary final : public IDictionary
     std::unique_ptr<ArenaWithFreeLists> string_arena;
 
     mutable std::exception_ptr last_exception;
-    mutable size_t error_count = 0;
-    mutable std::chrono::system_clock::time_point backoff_end_time;
+    mutable std::atomic<size_t> error_count = 0;
+    mutable std::atomic<std::chrono::system_clock::time_point> backoff_end_time{std::chrono::system_clock::time_point{}};
 
     mutable pcg64 rnd_engine;
 
     mutable size_t bytes_allocated = 0;
     mutable std::atomic<size_t> element_count{0};
     mutable std::atomic<size_t> hit_count{0};
     mutable std::atomic<size_t> query_count{0};
-};
+
+    /// Field and methods correlated with update expired and not found keys
+
+    using PresentIdHandler = std::function<void(Key, size_t)>;
+    using AbsentIdHandler  = std::function<void(Key, size_t)>;
+
+    struct UpdateUnit
```

Comment:
```suggestion
    /*
     * How the update goes: we basically have a method get(keys)->values. Values are cached, so sometimes we
     * can return them from the cache. For values not in cache, we query them from the dictionary, and add to the
     * cache. The cache is lossy, so we can't expect it to store all the keys, and we store them separately. Normally,
     * they would be passed as a return value of get(), but for Unknown Reasons the dictionaries use a baroque 
     * interface where get() accepts two callback, one that it calls for found values, and one for not found.
     *
     * Now we make it even uglier by doing this from multiple threads. The missing values are retreived from the
     * dictionary in a background thread, and this thread calls the provided callback. So if you provide the callbacks,
     * you MUST wait until the background update finishes, or god knows what happens. Unfortunately, we have no
     * way to check that you did this right, so good luck.
     */
    struct UpdateUnit
```

Reply:
Am I getting this right?

Reply:
Yes, I will commit suggestion and correct it by myself.